# Chapter 5 - Exercise 1: GroupBy - DONE

#### drinks.csv là tập tin cung cấp dữ liệu về tình hình tiêu thụ rượu bia ở các quốc gia theo từng châu lục

country: tên quốc gia tiêu thụ<br/>
beer_servings: số lượng bia tiêu thụ<br/>
spirit_servings: số lượng rượu mạnh (>= 40<sup>0</sup>)<br/>
wine_servings: số lượng rượu vang<br/>
total_litres_of_pure_alcohol: tổng số lít rượu nguyên chất<br/>
continent: châu lục

#### Thực hiện các yêu cầu sau, và đối chiếu với kết quả cho trước:

In [2]:
import pandas as pd

In [3]:
# Câu 1: Đọc dữ liệu từ tập tin drinks.csv với index_col là cột đầu tiên của dữ liệu, và lưu vào biến drink. 
drink = pd.read_csv('./drinks.csv', index_col = 0)
# Cho biết kiểu dữ liệu (type) của drink
print(type(drink))
# Cho biết ra kích thước (shape) của drink 
print(drink.shape)
# Hiển thị tên các cột (columns) của drink
print(drink.columns)

<class 'pandas.core.frame.DataFrame'>
(193, 6)
Index(['country', 'beer_servings', 'spirit_servings', 'wine_servings',
       'total_litres_of_pure_alcohol', 'continent'],
      dtype='object')


<details>
  <summary>Nhấn vào đây để xem kết quả!</summary>
  <pre>&lt;class 'pandas.core.frame.DataFrame'&gt;
(193, 6)
Index(['country', 'beer_servings', 'spirit_servings', 'wine_servings',
       'total_litres_of_pure_alcohol', 'continent'],
      dtype='object')
</pre>
  
</details>

In [3]:
# Xem 5 dòng dữ liệu đầu tiên (head) của drink
drink.head()

country  beer_servings  spirit_servings  wine_servings  \
0  Afghanistan              0                0              0   
1      Albania             89              132             54   
2      Algeria             25                0             14   
3      Andorra            245              138            312   
4       Angola            217               57             45   

   total_litres_of_pure_alcohol continent  
0                           0.0        AS  
1                           4.9        EU  
2                           0.7        AF  
3                          12.4        EU  
4                           5.9        AF

<details>
  <summary>Nhấn vào đây để xem kết quả!</summary>
  <div>
<style scoped="">
    .dataframe tbody tr th:only-of-type {
        vertical-align: middle;
    }

    .dataframe tbody tr th {
        vertical-align: top;
    }

    .dataframe thead th {
        text-align: right;
    }
</style>
<table border="1" class="dataframe">
  <thead>
    <tr style="text-align: right;">
      <th></th>
      <th>country</th>
      <th>beer_servings</th>
      <th>spirit_servings</th>
      <th>wine_servings</th>
      <th>total_litres_of_pure_alcohol</th>
      <th>continent</th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <th>0</th>
      <td>Afghanistan</td>
      <td>0</td>
      <td>0</td>
      <td>0</td>
      <td>0.0</td>
      <td>AS</td>
    </tr>
    <tr>
      <th>1</th>
      <td>Albania</td>
      <td>89</td>
      <td>132</td>
      <td>54</td>
      <td>4.9</td>
      <td>EU</td>
    </tr>
    <tr>
      <th>2</th>
      <td>Algeria</td>
      <td>25</td>
      <td>0</td>
      <td>14</td>
      <td>0.7</td>
      <td>AF</td>
    </tr>
    <tr>
      <th>3</th>
      <td>Andorra</td>
      <td>245</td>
      <td>138</td>
      <td>312</td>
      <td>12.4</td>
      <td>EU</td>
    </tr>
    <tr>
      <th>4</th>
      <td>Angola</td>
      <td>217</td>
      <td>57</td>
      <td>45</td>
      <td>5.9</td>
      <td>AF</td>
    </tr>
  </tbody>
</table>
</div>
  
</details>

In [4]:
# Xem 5 dòng dữ liệu cuối cùng (tail) của drink
drink.tail()

country  beer_servings  spirit_servings  wine_servings  \
188  Venezuela            333              100              3   
189    Vietnam            111                2              1   
190      Yemen              6                0              0   
191     Zambia             32               19              4   
192   Zimbabwe             64               18              4   

     total_litres_of_pure_alcohol continent  
188                           7.7        SA  
189                           2.0        AS  
190                           0.1        AS  
191                           2.5        AF  
192                           4.7        AF

<details>
  <summary>Nhấn vào đây để xem kết quả!</summary>
  <div>
<style scoped="">
    .dataframe tbody tr th:only-of-type {
        vertical-align: middle;
    }

    .dataframe tbody tr th {
        vertical-align: top;
    }

    .dataframe thead th {
        text-align: right;
    }
</style>
<table border="1" class="dataframe">
  <thead>
    <tr style="text-align: right;">
      <th></th>
      <th>country</th>
      <th>beer_servings</th>
      <th>spirit_servings</th>
      <th>wine_servings</th>
      <th>total_litres_of_pure_alcohol</th>
      <th>continent</th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <th>188</th>
      <td>Venezuela</td>
      <td>333</td>
      <td>100</td>
      <td>3</td>
      <td>7.7</td>
      <td>SA</td>
    </tr>
    <tr>
      <th>189</th>
      <td>Vietnam</td>
      <td>111</td>
      <td>2</td>
      <td>1</td>
      <td>2.0</td>
      <td>AS</td>
    </tr>
    <tr>
      <th>190</th>
      <td>Yemen</td>
      <td>6</td>
      <td>0</td>
      <td>0</td>
      <td>0.1</td>
      <td>AS</td>
    </tr>
    <tr>
      <th>191</th>
      <td>Zambia</td>
      <td>32</td>
      <td>19</td>
      <td>4</td>
      <td>2.5</td>
      <td>AF</td>
    </tr>
    <tr>
      <th>192</th>
      <td>Zimbabwe</td>
      <td>64</td>
      <td>18</td>
      <td>4</td>
      <td>4.7</td>
      <td>AF</td>
    </tr>
  </tbody>
</table>
</div>
  
</details>

In [5]:
# Câu 2: Cho biết số lượng bia tiêu thụ trung bình ở mỗi châu lục
drink.groupby('continent').mean()['beer_servings']

continent
AF     61.471698
AS     37.045455
EU    193.777778
OC     89.687500
SA    175.083333
Name: beer_servings, dtype: float64

<details>
  <summary>Nhấn vào đây để xem kết quả!</summary>
  <pre>continent
AF     61.471698
AS     37.045455
EU    193.777778
OC     89.687500
SA    175.083333
Name: beer_servings, dtype: float64</pre>
  
</details>

In [6]:
# Câu 3: Cho biết thông tin thống kê tổng quát (describe) số lượng rượu vang được tiêu thụ ở mỗi châu lục 
drink.groupby('continent')['wine_servings'].describe()

count        mean        std  min   25%    50%     75%    max
continent                                                               
AF          53.0   16.264151  38.846419  0.0   1.0    2.0   13.00  233.0
AS          44.0    9.068182  21.667034  0.0   0.0    1.0    8.00  123.0
EU          45.0  142.222222  97.421738  0.0  59.0  128.0  195.00  370.0
OC          16.0   35.625000  64.555790  0.0   1.0    8.5   23.25  212.0
SA          12.0   62.416667  88.620189  1.0   3.0   12.0   98.50  221.0

<details>
  <summary>Nhấn vào đây để xem kết quả!</summary>
  <div>
<style scoped="">
    .dataframe tbody tr th:only-of-type {
        vertical-align: middle;
    }

    .dataframe tbody tr th {
        vertical-align: top;
    }

    .dataframe thead th {
        text-align: right;
    }
</style>
<table border="1" class="dataframe">
  <thead>
    <tr style="text-align: right;">
      <th></th>
      <th>count</th>
      <th>mean</th>
      <th>std</th>
      <th>min</th>
      <th>25%</th>
      <th>50%</th>
      <th>75%</th>
      <th>max</th>
    </tr>
    <tr>
      <th>continent</th>
      <th></th>
      <th></th>
      <th></th>
      <th></th>
      <th></th>
      <th></th>
      <th></th>
      <th></th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <th>AF</th>
      <td>53.0</td>
      <td>16.264151</td>
      <td>38.846419</td>
      <td>0.0</td>
      <td>1.0</td>
      <td>2.0</td>
      <td>13.00</td>
      <td>233.0</td>
    </tr>
    <tr>
      <th>AS</th>
      <td>44.0</td>
      <td>9.068182</td>
      <td>21.667034</td>
      <td>0.0</td>
      <td>0.0</td>
      <td>1.0</td>
      <td>8.00</td>
      <td>123.0</td>
    </tr>
    <tr>
      <th>EU</th>
      <td>45.0</td>
      <td>142.222222</td>
      <td>97.421738</td>
      <td>0.0</td>
      <td>59.0</td>
      <td>128.0</td>
      <td>195.00</td>
      <td>370.0</td>
    </tr>
    <tr>
      <th>OC</th>
      <td>16.0</td>
      <td>35.625000</td>
      <td>64.555790</td>
      <td>0.0</td>
      <td>1.0</td>
      <td>8.5</td>
      <td>23.25</td>
      <td>212.0</td>
    </tr>
    <tr>
      <th>SA</th>
      <td>12.0</td>
      <td>62.416667</td>
      <td>88.620189</td>
      <td>1.0</td>
      <td>3.0</td>
      <td>12.0</td>
      <td>98.50</td>
      <td>221.0</td>
    </tr>
  </tbody>
</table>
</div>
  
</details>

In [13]:
# Câu 4: Cho biết số lượng các loại bia và rượu tiêu thụ trung bình (mean) ở mỗi châu lục
drink.groupby('continent').mean()

beer_servings  spirit_servings  wine_servings  \
continent                                                  
AF             61.471698        16.339623      16.264151   
AS             37.045455        60.840909       9.068182   
EU            193.777778       132.555556     142.222222   
OC             89.687500        58.437500      35.625000   
SA            175.083333       114.750000      62.416667   

           total_litres_of_pure_alcohol  
continent                                
AF                             3.007547  
AS                             2.170455  
EU                             8.617778  
OC                             3.381250  
SA                             6.308333

<details>
  <summary>Nhấn vào đây để xem kết quả!</summary>
  <div>
<style scoped="">
    .dataframe tbody tr th:only-of-type {
        vertical-align: middle;
    }

    .dataframe tbody tr th {
        vertical-align: top;
    }

    .dataframe thead th {
        text-align: right;
    }
</style>
<table border="1" class="dataframe">
  <thead>
    <tr style="text-align: right;">
      <th></th>
      <th>beer_servings</th>
      <th>spirit_servings</th>
      <th>wine_servings</th>
      <th>total_litres_of_pure_alcohol</th>
    </tr>
    <tr>
      <th>continent</th>
      <th></th>
      <th></th>
      <th></th>
      <th></th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <th>AF</th>
      <td>61.471698</td>
      <td>16.339623</td>
      <td>16.264151</td>
      <td>3.007547</td>
    </tr>
    <tr>
      <th>AS</th>
      <td>37.045455</td>
      <td>60.840909</td>
      <td>9.068182</td>
      <td>2.170455</td>
    </tr>
    <tr>
      <th>EU</th>
      <td>193.777778</td>
      <td>132.555556</td>
      <td>142.222222</td>
      <td>8.617778</td>
    </tr>
    <tr>
      <th>OC</th>
      <td>89.687500</td>
      <td>58.437500</td>
      <td>35.625000</td>
      <td>3.381250</td>
    </tr>
    <tr>
      <th>SA</th>
      <td>175.083333</td>
      <td>114.750000</td>
      <td>62.416667</td>
      <td>6.308333</td>
    </tr>
  </tbody>
</table>
</div>
  
</details>

In [14]:
# Câu 5: Cho biết giá trị trung vị (median) của các loại bia và rượu tiêu thụ ở mỗi châu lục
drink.groupby('continent').median()

beer_servings  spirit_servings  wine_servings  \
continent                                                  
AF                  32.0              3.0            2.0   
AS                  17.5             16.0            1.0   
EU                 219.0            122.0          128.0   
OC                  52.5             37.0            8.5   
SA                 162.5            108.5           12.0   

           total_litres_of_pure_alcohol  
continent                                
AF                                 2.30  
AS                                 1.20  
EU                                10.00  
OC                                 1.75  
SA                                 6.85

<details>
  <summary>Nhấn vào đây để xem kết quả!</summary>
  <div>
<style scoped="">
    .dataframe tbody tr th:only-of-type {
        vertical-align: middle;
    }

    .dataframe tbody tr th {
        vertical-align: top;
    }

    .dataframe thead th {
        text-align: right;
    }
</style>
<table border="1" class="dataframe">
  <thead>
    <tr style="text-align: right;">
      <th></th>
      <th>beer_servings</th>
      <th>spirit_servings</th>
      <th>wine_servings</th>
      <th>total_litres_of_pure_alcohol</th>
    </tr>
    <tr>
      <th>continent</th>
      <th></th>
      <th></th>
      <th></th>
      <th></th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <th>AF</th>
      <td>32.0</td>
      <td>3.0</td>
      <td>2.0</td>
      <td>2.30</td>
    </tr>
    <tr>
      <th>AS</th>
      <td>17.5</td>
      <td>16.0</td>
      <td>1.0</td>
      <td>1.20</td>
    </tr>
    <tr>
      <th>EU</th>
      <td>219.0</td>
      <td>122.0</td>
      <td>128.0</td>
      <td>10.00</td>
    </tr>
    <tr>
      <th>OC</th>
      <td>52.5</td>
      <td>37.0</td>
      <td>8.5</td>
      <td>1.75</td>
    </tr>
    <tr>
      <th>SA</th>
      <td>162.5</td>
      <td>108.5</td>
      <td>12.0</td>
      <td>6.85</td>
    </tr>
  </tbody>
</table>
</div>
  
</details>

In [21]:
# Câu 6: Cho biết số lượng rượu mạnh (spirit_servings) tiêu thụ trung bình, lớn nhất và nhỏ nhất ở mỗi châu lục 
# Gợi ý: Dùng spirit_servings.agg(['mean', 'min', 'max'])
spirit_servings = drink.groupby('continent')['spirit_servings']
spirit_servings.aggregate(['mean', 'min', 'max'])

mean  min  max
continent                      
AF          16.339623    0  152
AS          60.840909    0  326
EU         132.555556    0  373
OC          58.437500    0  254
SA         114.750000   25  302

<details>
  <summary>Nhấn vào đây để xem kết quả!</summary>
  <div>
<style scoped="">
    .dataframe tbody tr th:only-of-type {
        vertical-align: middle;
    }

    .dataframe tbody tr th {
        vertical-align: top;
    }

    .dataframe thead th {
        text-align: right;
    }
</style>
<table border="1" class="dataframe">
  <thead>
    <tr style="text-align: right;">
      <th></th>
      <th>mean</th>
      <th>min</th>
      <th>max</th>
    </tr>
    <tr>
      <th>continent</th>
      <th></th>
      <th></th>
      <th></th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <th>AF</th>
      <td>16.339623</td>
      <td>0</td>
      <td>152</td>
    </tr>
    <tr>
      <th>AS</th>
      <td>60.840909</td>
      <td>0</td>
      <td>326</td>
    </tr>
    <tr>
      <th>EU</th>
      <td>132.555556</td>
      <td>0</td>
      <td>373</td>
    </tr>
    <tr>
      <th>OC</th>
      <td>58.437500</td>
      <td>0</td>
      <td>254</td>
    </tr>
    <tr>
      <th>SA</th>
      <td>114.750000</td>
      <td>25</td>
      <td>302</td>
    </tr>
  </tbody>
</table>
</div>
  
</details>

In [31]:
# Câu 7: Sắp xếp dữ liệu tăng dần (sort_values) theo số lượng bia tiêu thụ
asc_sorted_drink = drink.sort_values('beer_servings')
sorted_drink

country  beer_servings  spirit_servings  wine_servings  \
117         Namibia            376                3              1   
45   Czech Republic            361              170            134   
62            Gabon            347               98             59   
65          Germany            346              117            175   
98        Lithuania            343              244             56   
..              ...            ...              ...            ...   
107      Mauritania              0                0              0   
158         Somalia              0                0              0   
111          Monaco              0                0              0   
128        Pakistan              0                0              0   
0       Afghanistan              0                0              0   

     total_litres_of_pure_alcohol continent  
117                           6.8        AF  
45                           11.8        EU  
62                            8.9        AF  
65                           11.3        EU  
98                           12.9        EU  
..                            ...       ...  
107                           0.0        AF  
158                           0.0        AF  
111                           0.0        EU  
128                           0.0        AS  
0                             0.0        AS  

[193 rows x 6 columns]

<details>
  <summary>Nhấn vào đây để xem kết quả!</summary>
  <div>
<style scoped="">
    .dataframe tbody tr th:only-of-type {
        vertical-align: middle;
    }

    .dataframe tbody tr th {
        vertical-align: top;
    }

    .dataframe thead th {
        text-align: right;
    }
</style>
<table border="1" class="dataframe">
  <thead>
    <tr style="text-align: right;">
      <th></th>
      <th>country</th>
      <th>beer_servings</th>
      <th>spirit_servings</th>
      <th>wine_servings</th>
      <th>total_litres_of_pure_alcohol</th>
      <th>continent</th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <th>0</th>
      <td>Afghanistan</td>
      <td>0</td>
      <td>0</td>
      <td>0</td>
      <td>0.0</td>
      <td>AS</td>
    </tr>
    <tr>
      <th>40</th>
      <td>Cook Islands</td>
      <td>0</td>
      <td>254</td>
      <td>74</td>
      <td>5.9</td>
      <td>OC</td>
    </tr>
    <tr>
      <th>79</th>
      <td>Iran</td>
      <td>0</td>
      <td>0</td>
      <td>0</td>
      <td>0.0</td>
      <td>AS</td>
    </tr>
    <tr>
      <th>90</th>
      <td>Kuwait</td>
      <td>0</td>
      <td>0</td>
      <td>0</td>
      <td>0.0</td>
      <td>AS</td>
    </tr>
    <tr>
      <th>97</th>
      <td>Libya</td>
      <td>0</td>
      <td>0</td>
      <td>0</td>
      <td>0.0</td>
      <td>AF</td>
    </tr>
    <tr>
      <th>...</th>
      <td>...</td>
      <td>...</td>
      <td>...</td>
      <td>...</td>
      <td>...</td>
      <td>...</td>
    </tr>
    <tr>
      <th>135</th>
      <td>Poland</td>
      <td>343</td>
      <td>215</td>
      <td>56</td>
      <td>10.9</td>
      <td>EU</td>
    </tr>
    <tr>
      <th>65</th>
      <td>Germany</td>
      <td>346</td>
      <td>117</td>
      <td>175</td>
      <td>11.3</td>
      <td>EU</td>
    </tr>
    <tr>
      <th>62</th>
      <td>Gabon</td>
      <td>347</td>
      <td>98</td>
      <td>59</td>
      <td>8.9</td>
      <td>AF</td>
    </tr>
    <tr>
      <th>45</th>
      <td>Czech Republic</td>
      <td>361</td>
      <td>170</td>
      <td>134</td>
      <td>11.8</td>
      <td>EU</td>
    </tr>
    <tr>
      <th>117</th>
      <td>Namibia</td>
      <td>376</td>
      <td>3</td>
      <td>1</td>
      <td>6.8</td>
      <td>AF</td>
    </tr>
  </tbody>
</table>
<p>193 rows × 6 columns</p>
</div>
  
</details>

In [35]:
# Cho biết 5 quốc gia có lượng tiêu thụ bia lớn nhất
desc_sort_drink = drink.sort_values('beer_servings', ascending = False)
desc_sort_drink.head()

country  beer_servings  spirit_servings  wine_servings  \
117         Namibia            376                3              1   
45   Czech Republic            361              170            134   
62            Gabon            347               98             59   
65          Germany            346              117            175   
98        Lithuania            343              244             56   

     total_litres_of_pure_alcohol continent  
117                           6.8        AF  
45                           11.8        EU  
62                            8.9        AF  
65                           11.3        EU  
98                           12.9        EU

<details>
  <summary>Nhấn vào đây để xem kết quả!</summary>
  <div>
<style scoped="">
    .dataframe tbody tr th:only-of-type {
        vertical-align: middle;
    }

    .dataframe tbody tr th {
        vertical-align: top;
    }

    .dataframe thead th {
        text-align: right;
    }
</style>
<table border="1" class="dataframe">
  <thead>
    <tr style="text-align: right;">
      <th></th>
      <th>country</th>
      <th>beer_servings</th>
      <th>spirit_servings</th>
      <th>wine_servings</th>
      <th>total_litres_of_pure_alcohol</th>
      <th>continent</th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <th>135</th>
      <td>Poland</td>
      <td>343</td>
      <td>215</td>
      <td>56</td>
      <td>10.9</td>
      <td>EU</td>
    </tr>
    <tr>
      <th>65</th>
      <td>Germany</td>
      <td>346</td>
      <td>117</td>
      <td>175</td>
      <td>11.3</td>
      <td>EU</td>
    </tr>
    <tr>
      <th>62</th>
      <td>Gabon</td>
      <td>347</td>
      <td>98</td>
      <td>59</td>
      <td>8.9</td>
      <td>AF</td>
    </tr>
    <tr>
      <th>45</th>
      <td>Czech Republic</td>
      <td>361</td>
      <td>170</td>
      <td>134</td>
      <td>11.8</td>
      <td>EU</td>
    </tr>
    <tr>
      <th>117</th>
      <td>Namibia</td>
      <td>376</td>
      <td>3</td>
      <td>1</td>
      <td>6.8</td>
      <td>AF</td>
    </tr>
  </tbody>
</table>
</div>
  
</details>

In [36]:
# Cho biết 5 quốc gia có lượng tiêu thụ bia nhỏ nhất
desc_sort_drink.tail()

country  beer_servings  spirit_servings  wine_servings  \
107   Mauritania              0                0              0   
158      Somalia              0                0              0   
111       Monaco              0                0              0   
128     Pakistan              0                0              0   
0    Afghanistan              0                0              0   

     total_litres_of_pure_alcohol continent  
107                           0.0        AF  
158                           0.0        AF  
111                           0.0        EU  
128                           0.0        AS  
0                             0.0        AS

<details>
  <summary>Nhấn vào đây để xem kết quả!</summary>
  <div>
<style scoped="">
    .dataframe tbody tr th:only-of-type {
        vertical-align: middle;
    }

    .dataframe tbody tr th {
        vertical-align: top;
    }

    .dataframe thead th {
        text-align: right;
    }
</style>
<table border="1" class="dataframe">
  <thead>
    <tr style="text-align: right;">
      <th></th>
      <th>country</th>
      <th>beer_servings</th>
      <th>spirit_servings</th>
      <th>wine_servings</th>
      <th>total_litres_of_pure_alcohol</th>
      <th>continent</th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <th>0</th>
      <td>Afghanistan</td>
      <td>0</td>
      <td>0</td>
      <td>0</td>
      <td>0.0</td>
      <td>AS</td>
    </tr>
    <tr>
      <th>40</th>
      <td>Cook Islands</td>
      <td>0</td>
      <td>254</td>
      <td>74</td>
      <td>5.9</td>
      <td>OC</td>
    </tr>
    <tr>
      <th>79</th>
      <td>Iran</td>
      <td>0</td>
      <td>0</td>
      <td>0</td>
      <td>0.0</td>
      <td>AS</td>
    </tr>
    <tr>
      <th>90</th>
      <td>Kuwait</td>
      <td>0</td>
      <td>0</td>
      <td>0</td>
      <td>0.0</td>
      <td>AS</td>
    </tr>
    <tr>
      <th>97</th>
      <td>Libya</td>
      <td>0</td>
      <td>0</td>
      <td>0</td>
      <td>0.0</td>
      <td>AF</td>
    </tr>
  </tbody>
</table>
</div>
  
</details>